commons-impl/src/main/java/com/connexta/transformation/commons/api/impl/AbstractMetadataImpl.java

```diff
@@ -0,0 +1,357 @@
+/**
+ * Copyright (c) Connexta
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.transformation.commons.api.impl;
+
+import com.connexta.transformation.commons.api.ErrorCode;
+import com.connexta.transformation.commons.api.MetadataTransformation;
+import com.connexta.transformation.commons.api.RequestInfo;
+import com.connexta.transformation.commons.api.exceptions.InvalidFieldException;
+import com.connexta.transformation.commons.api.exceptions.PersistenceException;
+import com.connexta.transformation.commons.api.exceptions.UnsupportedVersionException;
+import com.connexta.transformation.pojo.MetadataPojo;
+import com.connexta.transformation.pojo.TransformationPojo;
+import com.connexta.transformation.pojo.unknown.UnknownPojo;
+import com.google.common.annotations.VisibleForTesting;
+import io.micrometer.core.instrument.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.OptionalLong;
+import javax.annotation.Nullable;
+
+/**
+ * Provides an abstraction implementation for the {@link MetadataTransformation} interface which
+ * adds persistence support to be used by all concrete implementations.
+ */
+public abstract class AbstractMetadataImpl extends Persistable<MetadataPojo>
+    implements MetadataTransformation {
+  private static final String PERSISTABLE_TYPE = "metadata";
+
+  protected final Clock clock;
+
+  private String type;
+  private String transformId;
+  private RequestInfo requestInfo;
+  private Instant startTime;
+
+  protected volatile State state = State.IN_PROGRESS;
+  protected volatile ErrorCode failureReason;
+  protected volatile String failureMessage;
+  @Nullable protected volatile Instant completionTime = null;
+  @Nullable protected volatile String contentType = null;
+  protected volatile long contentLength = -1;
+
+  private boolean hasUnknowns = false;
+
+  /**
+   * Instantiates a new metadata with the specified information.
+   *
+   * @param type the type of metadata this structure will hold
+   * @param transformId the id of the transform request
+   * @param requestInfo the corresponding {@link RequestInfo} object
+   * @param clock the clock to use for retrieving wall and monotonic times
+   */
+  public AbstractMetadataImpl(
+      String type, String transformId, RequestInfo requestInfo, Clock clock) {
```

Comment:
Place holder for when we start calculating durations as we did in transform. On top of it, it makes testing much easier as you can mock that guy instead of relying on time.